docs/design_notes/jax_versioning.md

```diff
@@ -0,0 +1,175 @@
+# Jax and Jaxlib versioning
+
+## Why are `jax` and `jaxlib` separate packages?
+
+We publish JAX as two separate Python wheels, namely `jax`, which is a pure
+Python wheel, and `jaxlib`, which is a mostly-C++ wheel that contains libraries
+such as:
+* XLA,
+* pieces of LLVM used by XLA,
+* MLIR infrastructure, such as the MHLO Python bindings.
+* JAX-specific C++ libraries for fast JIT and PyTree manipulation.
+
+We distribute separate `jax` and `jaxlib` packages because it makes it easy to
+work on the Python parts of JAX without having to build C++ code or even having
+a C++ toolchain installed. `jaxlib` is a large library that is not easy for
+many users to build, but most changes to JAX only touch Python code. By
+allowing the Python pieces to be updated independently of the C++ pieces, we
+improve the development velocity for Python changes.
+
+In addition `jaxlib` is not cheap to build, but we want to be able to iterate on
+and run the JAX tests in environments without a lot of CPU, for example in
+Github Actions or on a laptop. Many of our CI builds simply use a prebuilt
+`jaxlib`, rather than needing to rebuild the C++ pieces of JAX on each PR.
+
+As we will see, distributing `jax` and `jaxlib` separately comes with a cost, in
+that it requires that changes to `jaxlib` maintain a backward compatible API.
+However, we believe that on balance it is preferable to make Python changes
+easy, even if at the cost of making C++ changes slightly harder.
+
+
+## How are `jax` and `jaxlib` versioned?
+
+Both `jax` and `jaxlib` releases are numbered `x.y.z`, where `x` is the major
+version, and `y` is the minor version, and `z` is an optional patch release.
+Version numbers must have a major and minor version and must follow
+[PEP 440](https://www.python.org/dev/peps/pep-0440/). Version number comparisons
+are lexicographic comparisons on tuples of integers.
+
+Each `jax` release has an associated minimum `jaxlib` version `mx.my.mz`. The
+minimum `jaxlib` version for `jax` version `x.y.z` must be no greater than
+`x.y.0`.
+
+For `jax` version `x.y.z` and `jaxlib` version `lx.ly.lz` to be compatible,
+the following must hold:
+
+* The jaxlib version (`lx.ly.lz`) must be greater than or equal to the minimum
+  jaxlib version (`mx.my.mz`).
+* The jax version (`x.y.z`) must have major and minor versions (i.e., `x.y`)
+  greater than or equal to the jaxlib major and minor versions (i.e., `mx.my`).
+  For example, `jax` 3.4.5 is compatible with `jaxlib` 3.4.9, but not `jaxlib`
+  3.5.0 or 4.1.3.
```

Comment:
Given that we very rarely bump the jax major or minor version numbers, will guaranteeing jax 0.2.X compatibility with future jaxlibs indefinitely be a problem? E.g. what about when we change APIs?

Reply:
Changed the design as discussed offline.